### ISSUE TITLE
Overlay operation

### ISSUE THREAD

**Issue by jorisvandenbossche** [pygeos/pygeos#135]
In addition to a spatial join, an overlay operation might also be interesting to see if this can be sped up in pygeos.

Referencing an old geopands PR here that has a C implementation: https://github.com/geopandas/geopandas/pull/479

**Comment by adriangb** [comment]
2 comments on this:

1) using qury_bulk should speed things up quite a bit
2) optimizing some of the other checks to be vectorized should also be possible
3) I'm not sure how much time it is taking, but there is a `reduce` over a `shapely.difference` being done in `_overlay_difference`. It should be possible to use [numpy's reduce](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ufunc.reduce.html).

**Comment by caspervdw** [comment]
If not strictly necessary, I would like to keep the API small. So if you can already perform all the heavy lifting using the current API, then I am :+1: on doing nothing.  Please let me know if you need anything.

**Comment by brendan-ward** [comment]
@jorisvandenbossche @adriangb is there a description of what the overlay operation is supposed to do?  I couldn't quite get my head around the Cython example; it wasn't immediately obvious what was overlay-related in that vs code to support use of the tree or otherwise ancillary.

**Comment by jorisvandenbossche** [comment]
I don't know directly of a detailed description, except of reading source code of geopandas, or qgis, or ..  There is also a high-level overview here: http://wiki.gis.com/wiki/index.php/Overlay 
But basically you can see it as a spatial join but instead of checking a certain predicate (eg `intersects`), you also calculate the actual set operation (eg `intersection`).

Note, it is certainly possible that all low-level tools to efficiently implement `overlay` are already available in pygeos, without the need to create a new custom overlay algorithm like the linked PR does. I didn't check that when opening the issue, it would require some experimentation. But I suppose that doing the STRtree bulk query and apply the set operation on the result  would already come close.

**Comment by jorisvandenbossche** [comment]
> doing the STRtree bulk query and apply the set operation on the result would already come close.

That's also basically what the geopandas python implementation is doing now, but using the manual looping with the rtree index instead of a bulk query:

https://github.com/geopandas/geopandas/blob/084d5848e4dde74e6915b8ab2274448bdf6e4392/geopandas/tools/overlay.py#L23-L44

**Comment by adriangb** [comment]
> > doing the STRtree bulk query and apply the set operation on the result would already come close.
> 
> That's also basically what the geopandas python implementation is doing now, but using the manual looping with the rtree index instead of a bulk query:

Yeah, that's what I was referring to above. I'm fairly certain that just by implementing `bulk_query` and vectorized predicate checking we'd get 90+% of the speedup.

**Comment by adriangb** [comment]
Just a quick update: from my testing in https://github.com/geopandas/geopandas/pull/1428#issuecomment-629704547, we have a lot of room for improvement on the geopandas side by vectorizing not only the tree queries and predicate checks but also the computation of the resultant geometries, empty results, etc.
